### ISSUE TITLE
Random segfaults (pointers to objects become invalid?)

### ISSUE THREAD

**Issue by dfdx** [aviks/JavaCall.jl#20]
Recently I got a new source for segmentation faults. This time, however, errors are not strictly reproducible, but somewhat random. Below is an example Julia session with actual code that I'm working on. In general, when done carefully, following sequence of commands does NOT produce any error, resulting in valid references to Java objects. Sometimes, however, things like in this session happen. 

First, I load my code by including facade file "core.jl":
```
julia> include("core.jl")
Loaded /usr/lib/jvm/jdk1.7.0_79/jre/lib/amd64/server/libjvm.so
demo (generic function with 1 method)
```

Then I create some objects with references to Java objects, and it works perfectly fine:
```
julia> sc = SparkContext()
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
15/08/08 01:26:16 INFO SparkContext: Running Spark version 1.3.1
...
SparkContext(JavaObject{symbol("org.apache.spark.api.java.JavaSparkContext")}(Ptr{Void} @0x000000000845c048))

julia> java_rdd = text_file(sc, "file:///var/log/syslog")
15/08/08 01:26:36 INFO MemoryStore: ensureFreeSpace(182921) called with curMem=0, maxMem=515553361
...
JavaRDD(JavaObject{symbol("org.apache.spark.api.java.JavaRDD")}(Ptr{Void} @0x000000000845c058))
```

But later code with references to just created objects may unexpectedly fail with `NullPointerException`:
```
julia> rdd = PipelinedRDD(java_rdd, identity)
Exception in thread "main" java.lang.NullPointerException
	at org.apache.spark.api.java.JavaRDD$.toRDD(JavaRDD.scala:197)
	at org.apache.spark.api.julia.JuliaRDD$.fromJavaRDD(JuliaRDD.scala:176)
	at org.apache.spark.api.julia.JuliaRDD.fromJavaRDD(JuliaRDD.scala)
ERROR: Error calling Java: java.lang.NullPointerException
 in error at error.jl:21
 in geterror at /home/<username>/.julia/v0.3/JavaCall/src/core.jl:206
 in _jcall at /home/<username>/.julia/v0.3/JavaCall/src/core.jl:174
 in jcall at /home/<username>/.julia/v0.3/JavaCall/src/core.jl:110
 in PipelinedRDD at /home/<username>/.julia/v0.3/Sparta/src/rdd.jl:19
```

And when trying to repeat calls that were valid 2 minutes ago result in a segmentation fault:
```
julia> java_rdd = text_file(sc, "file:///var/log/syslog")

signal (11): Segmentation fault
unknown function (ip: -947006856)
unknown function (ip: -946938247)
_jcall at /home/<username>/.julia/v0.3/JavaCall/src/core.jl:173
jl_apply_generic at /usr/local/julia03/bin/../lib/julia/libjulia.so (unknown line)
jl_f_apply at /usr/local/julia03/bin/../lib/julia/libjulia.so (unknown line)
jcall at /home/<username>/.julia/v0.3/JavaCall/src/core.jl:124
jl_apply_generic at /usr/local/julia03/bin/../lib/julia/libjulia.so (unknown line)
text_file at /home/<username>/.julia/v0.3/Sparta/src/context.jl:13
jl_apply_generic at /usr/local/julia03/bin/../lib/julia/libjulia.so (unknown line)
unknown function (ip: -508562872)
unknown function (ip: -508566343)
unknown function (ip: -508565685)
unknown function (ip: -508561350)
unknown function (ip: -508568005)
unknown function (ip: -508498911)
jl_f_top_eval at /usr/local/julia03/bin/../lib/julia/libjulia.so (unknown line)
eval_user_input at REPL.jl:53
jlcall_eval_user_input_20017 at  (unknown line)
jl_apply_generic at /usr/local/julia03/bin/../lib/julia/libjulia.so (unknown line)
anonymous at task.jl:95
jl_handle_stack_switch at /usr/local/julia03/bin/../lib/julia/libjulia.so (unknown line)
julia_trampoline at /usr/local/julia03/bin/../lib/julia/libjulia.so (unknown line)
unknown function (ip: 4199805)
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 4199861)
Segmentation fault
```

So it looks like references to some Java objects become invalid (because of garbage collection?) and all kinds of fun happen. 

I understand that an issue formulated in such a way is terribly hard to debug, so I'm ready to provide any needed information. Alternatively, you can reproduce it yourself using special branch (`segfault-demo`) in my repository (https://github.com/dfdx/Sparta.jl). To do so: 

1. Clone repository and checkout the branch: `git clone https://github.com/dfdx/Sparta.jl; cd Sparta; git checkout segfault-demo`
2. Build Java classes (assuming you have Maven installed): `cd mvn/sparta; mvn clean package`
3. Start Julia session and perform operations like above several times, may be with some pauses. 

I was able to reproduce it with Oracle and OpenJDK 1.7, on Julia 0.3 and 0.4.

**Comment by aviks** [comment]
Maybe we need to create `GlobalRef` for all objects? I didn't think we needed to... but maybe we do?? 

I'll try to replicate. `mvn package` will get me everything? I don't need to start a Spark process or anything?

**Comment by dfdx** [comment]
Good that you asked - I updated the branch so that now `mvn package` downloads all dependencies including Spark itself and packages them into a single "sparta-assembly.jar". This jar is then included in a classpath (see "init.jl"), and Spark context is created from Julia itself (see `sc = SparkContext()` above). So yes, now you only need to run `mvn package` to prepare all Java code.

**Comment by dfdx** [comment]
Regarding global references, yes, it looks pretty likely to be the cause. Would returning global refs from `jcall` and removing them in `deleteref` enough in this case?

**Comment by aviks** [comment]
Yes, that is what I was thinking. Convert object references to global refs
in `convert_result` and remove them in `deleteref`. Can you try that and
see if that helps?

**Comment by dfdx** [comment]
Just tried it in [global-refs](https://github.com/dfdx/JavaCall.jl/tree/global-refs) branch, but with no success. I wrapped all `convert_result` methods except for those for strings and arrays, so custom objects (like `SparkContext` or `JavaRDD`) should have been affected. Yet, I get same `NullPointerException` and segmentation fault.

**Comment by aviks** [comment]
:( The changes in the branch look right... 

`Exception in thread "main" java.lang.NullPointerException
    at org.apache.spark.api.java.JavaRDD$.toRDD(JavaRDD.scala:197)`

What is the object that is null in `JavaRDD.scala:197`? is it `sc` or `java_rdd`?

**Comment by dfdx** [comment]
It's `java_rdd`, or, more precisely, `java_rdd.jrdd`, because `java_rdd` is a Julia wrapper for a pointer to `JavaRDD` object in JVM. In source code referred method looks like this: 

    implicit def toRDD[T](rdd: JavaRDD[T]): RDD[T] = rdd.rdd

So basically it simply tries to get `.rdd` field of `JavaRDD` object. 

Anyway, I believe it's not essential here - I've just got segfault on `PipelineRDD` creation and I think I had `NullPointerException` when creating `SparkContext`.

**Comment by aviks** [comment]
fixed by #28
